### PR TITLE
Profile: Add pssn visibility toggle and consent update

### PR DESCRIPTION
**Pull request by GiuliaTeggi**
### Ticket:

- [14975](https://dev.azure.com/OGCIO-Digital-Services/Digital%20Services%20Programme/_workitems/edit/14975)

### Description

- Add `ppsn_visible` and `consent_to_prefill_data` columns to `user_details` table 
- Add toggling of ppsn field visibility
- Add button to save the 'consent to prefill data'

## Type

- [ ] **Dependency upgrade**
- [ ] **Bug fix**
- [ ] **New feature**
- [ ] **Dev change**

### Checklist:

- [ ] I have added tests that prove my fix is effective or that my feature works

### Screenshots:

N/A

### CODE REVIEW

File: apps/profile/app/components/UserDetails.tsx [hunk 1]
```diff
@@ -12,35 +120,29 @@ async function getUserDetails() {
...
-      phone: "01234567891",
-    };
-  }
+  }>(
+    `SELECT title, date_of_birth, ppsn, ppsn_visible, gender, phone FROM user_details WHERE user_id = $1`,
+    [userId],
+  );
+
+  const { title, date_of_birth, ppsn, ppsn_visible, gender, phone } =
+    res.rows[0] || {};
```
Comment (Thurfjell) [inline apps/profile/app/components/UserDetails.tsx:132]: rows.at(0) or rows?.[0] necessary?

File: apps/profile/migrations/sqls/20240405132138-init-up.sql [hunk 2]
```diff
@@ -1,13 +1,15 @@
 CREATE TABLE IF NOT EXISTS user_details(
     user_id UUID NOT NULL DEFAULT gen_random_uuid() UNIQUE,
-    title TEXT NOT NULL,
```
Comment (Thurfjell) [inline apps/profile/migrations/sqls/20240405132138-init-up.sql:3]: Im not sure we can do changes to a migration like this. You might have to create a new migration and alter table. Check with @williamwd

File: apps/profile/migrations/sqls/20240405132138-init-up.sql [hunk 3]
```diff
@@ -1,13 +1,15 @@
...
+    title TEXT,
     firstName TEXT NOT NULL,
     lastName TEXT NOT NULL,
-    date_of_birth DATE NOT NULL,
-    ppsn TEXT NOT NULL,
-    gender TEXT NOT NULL,
-    phone TEXT NOT NULL, 
+    date_of_birth DATE,
```
Comment (Thurfjell) [inline apps/profile/migrations/sqls/20240405132138-init-up.sql:6]: timestamptz?

File: apps/profile/app/components/Consent.tsx [hunk 4]
```diff
@@ -1,16 +1,75 @@
...
+    WHERE user_id = $1
+    `,
+    [userId],
+  );
+
+  if (userExistsQuery.rows.length > 0) {
+    await postgres.pgpool.query(
+      `
+          UPDATE user_details
+          SET consent_to_prefill_data = $1
```
[the remark this answers is not in the document]
Comment (GiuliaTeggi) [reply apps/profile/app/components/Consent.tsx:26]: Yes, actually the wording is 'I consent to using the above data to pre-fill forms', not about our right to keep personal data.

File: apps/profile/app/components/Consent.tsx [hunk 5]
```diff
@@ -1,16 +1,75 @@
...
+async function getConsentData() {
+  const { userId } = await PgSessions.get();
+
+  const res = await postgres.pgpool.query<{
+    consent_to_prefill_data: boolean;
+  }>(`SELECT consent_to_prefill_data FROM user_details WHERE user_id = $1`, [
+    userId,
+  ]);
+
+  const { consent_to_prefill_data } = res.rows[0] || {};
```
[the remark this answers is not in the document]
Comment (GiuliaTeggi) [reply apps/profile/app/components/Consent.tsx:51]: No, res.rows is always defined, would just be an empty array, so res.rows[0] in that case does return `undefined`.

File: apps/profile/app/components/UserDetails.tsx (hunk 1 above)
Comment (GiuliaTeggi) [reply apps/profile/app/components/UserDetails.tsx:132]: As above, tested this, rows is just an empty array, so doing res.rows[0] returns `undefined`

File: apps/profile/migrations/sqls/20240405132138-init-up.sql (hunk 2 above)
Comment (GiuliaTeggi) [reply apps/profile/migrations/sqls/20240405132138-init-up.sql:3]: Yes will write a new migration

File: apps/profile/app/components/Consent.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (Thurfjell) [reply apps/profile/app/components/Consent.tsx:26]: Alright, so I guess we run two consents then. One upon registration (can we store your user data from gov id login) and then one in here "can we use this to pre-fill".

This seems kind of confusing? You have to opt in to the pre-fill one?

**Review by Thurfjell — APPROVED** [review]

File: apps/profile/app/components/Consent.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (GiuliaTeggi) [reply apps/profile/app/components/Consent.tsx:26]: I'll speak with Tiago about this, I am not sure how the user consent to store personal data is going to work to be honest.

**Pull request merged**